File: packages/occlude/src/plan.ts
/**
 * The ordered drawing plan as a value, and selections of it.
 *
 * The engine resolves the drawing once — visibility, finishing, merge,
 * tour, bridge — and encodes the result as one plan buffer with native
 * primitives (`wasm_plan`; layout in crates/occlude-core/src/plan.rs). This
 * module reads that buffer, gives it a content identity, and selects
 * contiguous chain ranges of it. Selection never re-plans: it neither
 * reorders, reverses, merges nor bridges, and it never reruns visibility —
 * dropping later ink does not reveal what it hid.
 *
 * Two questions about time are two operations: `selectTime` cuts an
 * interval of the FULL drawing's timeline (quantized to whole chains);
 * `fitDuration` finds the longest prefix of a selection whose STANDALONE
 * estimate fits a budget — a middle interval pays its own travel-in and
 * final lift, so it is not a difference of two timestamps.
 */

import { schedulePlan, type EstimateOpts, type PenTiming, type PlanEstimate, type PlanSchedule } from './motion.js';
import type { Prim } from './prims.js';

export const PLAN_SCHEMA = 1;

/** The path-optimization inputs of a plan — the only knobs planning has.
 * `optimize`: the 2-opt tour budget (`false` = nearest-neighbour order
 * only, a number overrides, default 200 000). `bridge`: draw through
 * sub-nib gaps instead of lifting — `false` never, a number is the gap
 * in mm for every pen, default half the nib per pen. A sketch states
 * them with `t.plan({...})`; `plan(result, opts)` takes them directly. */
export interface PlanOptions {
  optimize?: boolean | number;
  bridge?: boolean | number;
}

/** What part of the ordered plan a sketch asks to be drawn — code, so
 * the same program and seed give the same ink. Exactly one range form:
 * `chains` (whole chains, half-open), `progress` (a FRACTION OF CHAINS,
 * not ink or time), or `minutes` (an interval of the full plan's
 * estimated timeline, quantized to completed chains — needs a machine's
 * timing). `budget` (minutes) then keeps the longest prefix of that
 * range whose standalone estimate fits. Empty = the whole plan. */
export interface DrawRequest {
  chains?: [number, number];
  progress?: [number, number];
  minutes?: [number, number];
  budget?: number;
}

const pair = (v: unknown, what: string): [number, number] => {
  if (!Array.isArray(v) || v.length !== 2 || !v.every((x) => typeof x === 'number' && Number.isFinite(x))) throw new Error(`draw: ${what} must be [from, to] finite numbers`);
  if (v[0] > v[1]) throw new Error(`draw: ${what} from ${v[0]} exceeds to ${v[1]}`);
  return [v[0], v[1]];
};

/** Validate a sketch's request; the same object comes back normalized. */
export function checkDrawRequest(req: DrawRequest): DrawRequest {
  if (typeof req !== 'object' || req === null) throw new Error('draw: expected { chains | progress | minutes, budget? }');
  const forms = (['chains', 'progress', 'minutes'] as const).filter((k) => req[k] !== undefined);
  if (forms.length > 1) throw new Error(`draw: give one of chains, progress or minutes, not ${forms.join(' and ')}`);
  const out: DrawRequest = {};
  if (req.chains) {
    const c = pair(req.chains, 'chains');
    if (!Number.isInteger(c[0]) || !Number.isInteger(c[1]) || c[0] < 0) throw new Error('draw: chains must be non-negative integers');
    out.chains = c;
  }
  if (req.progress) {
    const p = pair(req.progress, 'progress');
    if (p[0] < 0 || p[1] > 1) throw new Error('draw: progress must lie within [0, 1]');
    out.progress = p;
  }
  if (req.minutes) {
    const m = pair(req.minutes, 'minutes');
    if (m[0] < 0) throw new Error('draw: minutes must be non-negative');
    out.minutes = m;
  }
  if (req.budget !== undefined) {
    if (typeof req.budget !== 'number' || !Number.isFinite(req.budget) || req.budget < 0) throw new Error('draw: budget must be a non-negative number of minutes');
    out.budget = req.budget;
  }
  for (const k of Object.keys(req)) if (!['chains', 'progress', 'minutes', 'budget'].includes(k)) throw new Error(`draw: unknown option '${k}'`);
  return out;
}

/** One pen-down run of the plan: native primitives, plan order. */
export interface PlanChain {
  /** Row in the FULL plan — the identity every consumer shares. */
  index: number;
  pen: number;
  dot: boolean;
  prims: Prim[];
}

/** What is needed to interpret and identify a plan besides its geometry. */
export interface PlanSettings {
  tourBudget: number;
  pens: { name: string; width: number }[];
  paper: { w: number; h: number };
  /** Sub-nib gap per pen that was drawn through instead of lifted. */
  bridgeGapMm: number[];
  /** Engine identity (build stamp or wasm digest) when the caller has one. */
  engine?: string;
}

export interface DrawingPlan {
  schemaVersion: number;
  /** SHA-256 over the canonical settings and the exact plan bytes. */
  planHash: string;
  chains: PlanChain[];
  settings: PlanSettings;
  /** The exact bytes the engine produced — what exports and saves use. */
  buffer: Float64Array;
}

/** A contiguous, half-open range `[fromChain, toChain)` of ONE exact plan. */
export interface PlanSelection {
  sourcePlanHash: string;
  fromChain: number;
  toChain: number;
  /** `${planHash}:${from}-${to}` — exact, readable, collision-free. */
  selectionHash: string;
  /** Chains selected — so nearby slider values that pick the same range are visible. */
  count: number;
}

// ---- buffer ---------------------------------------------------------------------

export function decodePlanBuffer(buf: Float64Array | number[]): PlanChain[] {
  let i = 0;
  const need = (n: number) => {
    if (i + n > buf.length) throw new Error('plan buffer truncated');
  };
  need(2);
  const schema = buf[i++];
  if (schema !== PLAN_SCHEMA) throw new Error(`plan schema ${schema} is not ${PLAN_SCHEMA}`);
  const n = buf[i++];
  const chains: PlanChain[] = [];
  for (let c = 0; c < n; c++) {
    need(3);
    const pen = buf[i++];
    const dot = buf[i++] !== 0;
    const np = buf[i++];
    const prims: Prim[] = [];
    for (let k = 0; k < np; k++) {
      need(1);
      const kind = buf[i++];
      if (kind === 0) {
        need(4);
        prims.push({ t: 'line', x0: buf[i], y0: buf[i + 1], x1: buf[i + 2], y1: buf[i + 3] });
        i += 4;
      } else if (kind === 1) {
        need(5);
        prims.push({ t: 'arc', cx: buf[i], cy: buf[i + 1], r: buf[i + 2], start: buf[i + 3], sweep: buf[i + 4] });
        i += 5;
      } else if (kind === 2) {
        need(8);
        prims.push({ t: 'cubic', x0: buf[i], y0: buf[i + 1], c0x: buf[i + 2], c0y: buf[i + 3], c1x: buf[i + 4], c1y: buf[i + 5], x1: buf[i + 6], y1: buf[i + 7] });
        i += 8;
      } else throw new Error(`plan buffer: unknown primitive kind ${kind}`);
    }
    if (prims.length === 0) throw new Error('plan buffer: a chain with no primitives');
    chains.push({ index: c, pen, dot, prims });
  }
  if (i !== buf.length) throw new Error('plan buffer: trailing data');
  return chains;
}

/** The inverse of `decodePlanBuffer` — the same bytes the engine writes, so
 * a saved selection is a plan buffer in its own right. */
export function encodePlanBuffer(chains: readonly PlanChain[]): Float64Array {
  const out: number[] = [PLAN_SCHEMA, chains.length];
  for (const c of chains) {
    out.push(c.pen, c.dot ? 1 : 0, c.prims.length);
    for (const p of c.prims) {
      if (p.t === 'line') out.push(0, p.x0, p.y0, p.x1, p.y1);
      else if (p.t === 'arc') out.push(1, p.cx, p.cy, p.r, p.start, p.sweep);
      else out.push(2, p.x0, p.y0, p.c0x, p.c0y, p.c1x, p.c1y, p.x1, p.y1);
    }
  }
  return Float64Array.from(out);
}

/** JSON with keys sorted at every level: one spelling per value. */
export function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== 'object') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(',')}]`;
  const o = value as Record<string, unknown>;
  return `{${Object.keys(o).sort().map((k) => `${JSON.stringify(k)}:${canonicalJson(o[k])}`).join(',')}}`;
}

const hex = (bytes: ArrayBuffer): string => Array.from(new Uint8Array(bytes), (b) => b.toString(16).padStart(2, '0')).join('');

/** Content identity: SHA-256 of `schema | canonical settings | plan bytes`
 * (little-endian float64). No timestamps, no object identities. */
export async function hashPlan(buffer: Float64Array, settings: PlanSettings): Promise<string> {
  const head = new TextEncoder().encode(`occlude-plan/${PLAN_SCHEMA}\n${canonicalJson(settings)}\n`);
  const body = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength);
  const all = new Uint8Array(head.length + body.length);
  all.set(head);
  all.set(body, head.length);
  return hex(await crypto.subtle.digest('SHA-256', all));
}

/** A plan value from the engine's buffer and the settings that made it. */
export async function makePlan(buffer: Float64Array, settings: PlanSettings): Promise<DrawingPlan> {
  const chains = decodePlanBuffer(buffer);
  const planHash = await hashPlan(buffer, settings);
  return { schemaVersion: PLAN_SCHEMA, planHash, chains, settings, buffer };
}

/** Rebuild a plan from saved bytes and verify its identity. */
export async function openPlan(buffer: Float64Array, settings: PlanSettings, expectedHash: string): Promise<DrawingPlan> {
  const plan = await makePlan(buffer, settings);
  if (plan.planHash !== expectedHash) throw new Error(`plan hash mismatch: saved ${expectedHash.slice(0, 12)}…, content ${plan.planHash.slice(0, 12)}…`);
  return plan;
}

// ---- sampled representation ---------------------------------------------------------

/** A chain flattened for machine execution and timing: `[pen, dot, n, x0, y0, …]`
 * per chain is the toolpath layout (`wasm_plan_toolpath`). */
export interface FlatChain {
  index: number;
  pen: number;
  dot: boolean;
  pts: Float64Array;
}

/** The ONE parser of the toolpath layout. `firstIndex` is the full-plan row
 * of the first chain, so a range's chains keep their source indices. */
export function parseToolpath(plan: Float64Array, firstIndex = 0): FlatChain[] {
  const chains: FlatChain[] = [];
  for (let i = 0; i < plan.length; ) {
    const pen = plan[i++];
    const dot = plan[i++] === 1;
    const n = plan[i++];
    chains.push({ index: firstIndex + chains.length, pen, dot, pts: plan.subarray(i, i + n * 2) });
    i += n * 2;
  }
  return chains;
}

/** Back to the toolpath layout — for drivers that take the buffer. */
export function encodeToolpath(chains: readonly FlatChain[]): Float64Array {
  let size = 0;
  for (const c of chains) size += 3 + c.pts.length;
  const out = new Float64Array(size);
  let i = 0;
  for (const c of chains) {
    out[i++] = c.pen;
    out[i++] = c.dot ? 1 : 0;
    out[i++] = c.pts.length / 2;
    out.set(c.pts, i);
    i += c.pts.length;
  }
  return out;
}

// ---- selections -------------------------------------------------------------------

function selection(plan: DrawingPlan, from: number, to: number): PlanSelection {
  return { sourcePlanHash: plan.planHash, fromChain: from, toChain: to, selectionHash: `${plan.planHash}:${from}-${to}`, count: to - from };
}

const finite = (v: number, what: string) => {
  if (typeof v !== 'number' || !Number.isFinite(v)) throw new Error(`${what} must be a finite number`);
};

/** Whole chains `[from, to)` of the full plan. */
export function selectChains(plan: DrawingPlan, range: { from: number; to: number }): PlanSelection {
  const n = plan.chains.length;
  finite(range.from, 'from');
  finite(range.to, 'to');
  if (!Number.isInteger(range.from) || !Number.isInteger(range.to)) throw new Error('chain indices must be integers');
  if (range.from < 0 || range.to > n || range.from > range.to) throw new Error(`chain range [${range.from}, ${range.to}) is not within 0..${n} in order`);
  return selection(plan, range.from, range.to);
}

/** The full plan `[0, N)`. */
export function selectAll(plan: DrawingPlan): PlanSelection {
  return selection(plan, 0, plan.chains.length);
}

/** By FRACTION OF CHAINS (not ink, area or time): boundaries resolve to
 * `floor(f · N)`, and 1 resolves to N. */
export function selectProgress(plan: DrawingPlan, range: { from: number; to: number }): PlanSelection {
  finite(range.from, 'from');
  finite(range.to, 'to');
  if (range.from < 0 || range.to > 1 || range.from > range.to) throw new Error(`progress range [${range.from}, ${range.to}) is not within 0..1 in order`);
  const n = plan.chains.length;
  const at = (f: number) => (f >= 1 ? n : Math.floor(f * n));
  return selection(plan, at(range.from), at(range.to));
}

/** The plan's timeline through the shared estimator. */
export function planSchedule(chains: readonly FlatChain[], penOf: (pen: number) => PenTiming | undefined, timing: EstimateOpts): PlanSchedule {
  return schedulePlan(chains, penOf, timing);
}

export interface TimeSelection {
  selection: PlanSelection;
  /** Requested interval, ms of the FULL plan's timeline. */
  requested: { fromMs: number; toMs: number };
  /** Where the boundaries actually fell: completion times of the last chain
   * before `from` and of the last chain in the range (0 / 0 when empty). */
  effective: { fromMs: number; toMs: number };
}

/** An interval of the FULL drawing's timeline, quantized to completed
 * chains: with B(t) = chains fully completed by time t, the range is
 * `[B(fromMs), B(toMs))`. Ties include every chain completing exactly at
 * t; `fromMs === 0` is special-cased to 0 so zero-duration chains at the
 * start are never lost. A chain the interval ends inside is excluded. The
 * start may resolve EARLIER than asked — quantization, not clipping. */
export function selectTime(
  plan: DrawingPlan,
  flat: readonly FlatChain[],
  range: { fromMs: number; toMs: number },
  schedule: PlanSchedule,
): TimeSelection {
  finite(range.fromMs, 'fromMs');
  finite(range.toMs, 'toMs');
  if (flat.length !== plan.chains.length) throw new Error('selectTime: the toolpath does not cover the whole plan');
  const total = schedule.estimate.totalMs;
  if (range.fromMs < 0 || range.toMs > total + 1e-9 || range.fromMs > range.toMs) {
    throw new Error(`time range [${range.fromMs}, ${range.toMs}) ms is not within 0..${Math.round(total)} in order`);
  }
  const done = schedule.chainStartMs.map((s, i) => s + schedule.chainDurMs[i]);
  const completedBy = (t: number) => {
    let k = 0;
    while (k < done.length && done[k] <= t + 1e-9) k++;
    return k;
  };
  const from = range.fromMs === 0 ? 0 : completedBy(range.fromMs);
  const to = completedBy(range.toMs);
  const sel = selection(plan, from, to);
  return {
    selection: sel,
    requested: { fromMs: range.fromMs, toMs: range.toMs },
    effective: { fromMs: from > 0 ? done[from - 1] : 0, toMs: to > 0 ? done[to - 1] : 0 },
  };
}

/** The flattened chains of a selection, source indices kept. */
export function selectedFlat(flat: readonly FlatChain[], sel: PlanSelection): FlatChain[] {
  return flat.slice(sel.fromChain, sel.toChain);
}

/** Estimated time to execute the selection ON ITS OWN: travel from the
 * origin to its first chain, the plan's travel between its chains, a
 * final rise to full lift. Not a difference of full-plan timestamps. */
export function standaloneEstimate(flat: readonly FlatChain[], sel: PlanSelection, penOf: (pen: number) => PenTiming | undefined, timing: EstimateOpts): PlanEstimate {
  return schedulePlan(selectedFlat(flat, sel), penOf, timing).estimate;
}

export interface FitResult {
  selection: PlanSelection;
  /** Standalone estimate of the fitted prefix, ms. */
  estimatedMs: number;
  unusedMs: number;
  /** Chains dropped from the end of the selection to fit. */
  dropped: number;
}

/** The longest prefix of `sel` (starting at its first chain, never
 * elsewhere) whose standalone estimate fits `budgetMs`. Priced from ONE
 * schedule of the selection: a prefix of k chains costs the schedule up to
 * chain k−1 with that chain's up-settle re-priced at full lift (the
 * terminal rule), which is exactly what a fresh estimate of the prefix
 * gives — verified against direct estimates in the tests. O(points) once
 * plus O(k) for the scan; no monotonicity assumed. Budget 0, or a first
 * chain that alone exceeds the budget, gives an empty selection. */
export function fitDuration(
  plan: DrawingPlan,
  flat: readonly FlatChain[],
  sel: PlanSelection,
  opts: { budgetMs: number },
  penOf: (pen: number) => PenTiming | undefined,
  timing: EstimateOpts,
): FitResult {
  finite(opts.budgetMs, 'budgetMs');
  if (opts.budgetMs < 0) throw new Error('budgetMs must be non-negative');
  if (sel.sourcePlanHash !== plan.planHash) throw new Error('fitDuration: the selection belongs to another plan');
  const mine = selectedFlat(flat, sel);
  const sched = schedulePlan(mine, penOf, timing);
  let best = 0;
  let bestMs = 0;
  for (let k = 1; k <= mine.length; k++) {
    const i = k - 1;
    const ms = sched.chainStartMs[i] + sched.chainDurMs[i] - sched.chainUpMs[i] + sched.chainUpFullMs[i];
    if (ms <= opts.budgetMs + 1e-9) {
      best = k;
      bestMs = ms;
    }
  }
  const fitted = selection(plan, sel.fromChain, sel.fromChain + best);
  return { selection: fitted, estimatedMs: bestMs, unusedMs: opts.budgetMs - bestMs, dropped: sel.count - best };
}

/** Timing a request by minutes or budget needs: the plan's sampled
 * chains and the machine's estimator inputs. */
export interface DrawTiming {
  flat: readonly FlatChain[];
  penOf: (pen: number) => PenTiming | undefined;
  opts: EstimateOpts;
}

export interface ResolvedDraw {
  request: DrawRequest;
  /** Before any budget fitting. */
  selection: PlanSelection;
  effective?: { fromMs: number; toMs: number };
  fit?: FitResult & { budgetMs: number };
  /** The selection every consumer uses: fitted when a budget was given. */
  final: PlanSelection;
  /** Standalone estimate of `final` and the whole plan's, when timed. */
  estimate?: PlanEstimate;
  fullMs?: number;
}

/** Resolve a sketch's `draw` request against a plan. Chain and progress
 * forms need nothing else; minutes and budget need `timing`, and throw
 * without it (the machine's clock is not the sketch's to know). A chain
 * range past the plan's end clamps; the plan is never changed. */
export function resolveDraw(plan: DrawingPlan, req: DrawRequest | undefined, timing?: DrawTiming): ResolvedDraw {
  const r = checkDrawRequest(req ?? {});
  const n = plan.chains.length;
  const needsTime = r.minutes !== undefined || r.budget !== undefined;
  if (needsTime && !timing) throw new Error('draw: a range in minutes or a budget needs the machine timing (pass timing, or choose chains / progress)');
  if (timing && timing.flat.length !== n) throw new Error('draw: the toolpath does not cover the whole plan');
  let selection: PlanSelection;
  let effective: ResolvedDraw['effective'];
  let schedule: PlanSchedule | undefined;
  const scheduled = () => (schedule ??= schedulePlan(timing!.flat, timing!.penOf, timing!.opts));
  if (r.chains) selection = selectChains(plan, { from: Math.min(r.chains[0], n), to: Math.min(r.chains[1], n) });
  else if (r.progress) selection = selectProgress(plan, { from: r.progress[0], to: r.progress[1] });
  else if (r.minutes) {
    const total = scheduled().estimate.totalMs;
    const t = selectTime(plan, timing!.flat, { fromMs: Math.min(r.minutes[0] * 60_000, total), toMs: Math.min(r.minutes[1] * 60_000, total) }, scheduled());
    selection = t.selection;
    effective = t.effective;
  } else selection = selectAll(plan);
  let fit: ResolvedDraw['fit'];
  if (r.budget !== undefined) {
    const budgetMs = r.budget * 60_000;
    fit = { ...fitDuration(plan, timing!.flat, selection, { budgetMs }, timing!.penOf, timing!.opts), budgetMs };
  }
  const final = fit?.selection ?? selection;
  const out: ResolvedDraw = { request: r, selection, effective, fit, final };
  if (timing) {
    out.estimate = standaloneEstimate(timing.flat, final, timing.penOf, timing.opts);
    out.fullMs = scheduled().estimate.totalMs;
  }
  return out;
}

/** A plan value from bytes whose identity is already known or not needed
 * (an export inside one process): the sync half of `makePlan`. */
export function planValue(buffer: Float64Array, settings: PlanSettings, planHash: string): DrawingPlan {
  return { schemaVersion: PLAN_SCHEMA, planHash, chains: decodePlanBuffer(buffer), settings, buffer };
}

/** Full-plan bounds of a set of flattened chains (paper mm). */
export function chainsBounds(chains: readonly FlatChain[]): { x: number; y: number; w: number; h: number } {
  let x0 = Infinity; let y0 = Infinity; let x1 = -Infinity; let y1 = -Infinity;
  for (const c of chains) {
    for (let k = 0; k < c.pts.length; k += 2) {
      const x = c.pts[k];
      const y = c.pts[k + 1];
      if (x < x0) x0 = x;
      if (y < y0) y0 = y;
      if (x > x1) x1 = x;
      if (y > y1) y1 = y;
    }
  }
  if (!Number.isFinite(x0)) return { x: 0, y: 0, w: 0, h: 0 };
  return { x: x0, y: y0, w: x1 - x0, h: y1 - y0 };
}

File: crates/occlude-core/src/plan.rs
//! The ordered drawing plan — ONE resolved merge → tour → bridge that every
//! consumer (SVG, G-code, toolpath, simulation, plotting, estimates) reads
//! instead of planning again.
//!
//! Order of operations: solve the full drawing (visibility, finishing) →
//! merge, tour and bridge per pen → materialize this plan → select a range
//! of it → derive representations. Selection never re-plans: a range is a
//! contiguous slice of the same chains in the same order and direction.
//!
//! The canonical timeline is pen by pen, in pen index order — that IS the
//! machine's actual order (there is no pen changer: one pen per run).
//!
//! Wire format (`encode_plan`), one `Vec<f64>`, exact and lossless:
//!   [schema = 1, n_chains,
//!    per chain: pen, dot(0|1), n_prims,
//!      per prim: kind, params…] with
//!      kind 0 = Line   x0 y0 x1 y1
//!      kind 1 = Arc    cx cy r start sweep
//!      kind 2 = Cubic  x0 y0 c0x c0y c1x c1y x1 y1
//! Native arcs and cubics survive, so the exact SVG/G-code paths keep them.

use crate::fragment::Frag;
use crate::gcode::{merge_chains, tour, Chain};
use crate::pipeline::Pen;
use crate::primitive::{Arc, Cubic, Line, Primitive};
use crate::route::bridge_chains;
use crate::vec2::{v, Vec2};

pub const PLAN_SCHEMA: f64 = 1.0;

/// Sub-nib gap a pen draws through instead of lifting, by default: half
/// the nib. `bridge` overrides it for every pen: `Some(0.0)` never
/// bridges, `Some(g)` bridges gaps up to `g` mm, `None` is the default.
pub fn bridge_gap(pen: &Pen, bridge: Option<f64>) -> f64 {
    match bridge {
        Some(g) => g.max(0.0),
        None => pen.width.max(0.05) * 0.5,
    }
}

/// The path-optimization inputs of a plan — the only knobs planning has.
#[derive(Debug, Clone, Copy)]
pub struct PlanOptions {
    /// 2-opt iteration budget for the tour; 0 keeps nearest-neighbour order.
    pub tour_budget: usize,
    /// Bridge gap override, see `bridge_gap`.
    pub bridge: Option<f64>,
}

impl Default for PlanOptions {
    fn default() -> PlanOptions {
        PlanOptions { tour_budget: 200_000, bridge: None }
    }
}

/// The full plan: for each pen in index order, merge → tour → bridge.
pub fn plan_chains(frags: &[Frag], pens: &[Pen], tour_budget: usize) -> Vec<Chain> {
    plan_chains_with(frags, pens, PlanOptions { tour_budget, bridge: None })
}

pub fn plan_chains_with(frags: &[Frag], pens: &[Pen], opts: PlanOptions) -> Vec<Chain> {
    let mut out: Vec<Chain> = Vec::new();
    for (pi, pen) in pens.iter().enumerate() {
        let chains = {
            let _z = crate::profile::zone("p1 merge");
            merge_chains(frags, pi as u32)
        };
        if chains.is_empty() {
            continue;
        }
        let chains = {
            let _z = crate::profile::zone("p2 tour");
            tour(chains, opts.tour_budget)
        };
        let _z = crate::profile::zone("p3 bridge");
        out.extend(bridge_chains(chains, bridge_gap(pen, opts.bridge)));
    }
    out
}

pub fn encode_plan(chains: &[Chain]) -> Vec<f64> {
    let _z = crate::profile::zone("p4 encode");
    let mut out: Vec<f64> = vec![PLAN_SCHEMA, chains.len() as f64];
    for c in chains {
        out.push(c.pen as f64);
        out.push(if c.dot { 1.0 } else { 0.0 });
        out.push(c.prims.len() as f64);
        for p in &c.prims {
            match p {
                Primitive::Line(l) => out.extend([0.0, l.p0.x, l.p0.y, l.p1.x, l.p1.y]),
                Primitive::Arc(a) => out.extend([1.0, a.center.x, a.center.y, a.r, a.start, a.sweep]),
                Primitive::Cubic(k) => out.extend([
                    2.0, k.p0.x, k.p0.y, k.c0.x, k.c0.y, k.c1.x, k.c1.y, k.p1.x, k.p1.y,
                ]),
            }
        }
    }
    out
}

pub fn decode_plan(buf: &[f64]) -> Result<Vec<Chain>, String> {
    let mut i = 0usize;
    let mut take = |n: usize| -> Result<&[f64], String> {
        if i + n > buf.len() {
            return Err("plan buffer truncated".into());
        }
        let s = &buf[i..i + n];
        i += n;
        Ok(s)
    };
    let head = take(2)?;
    if head[0] != PLAN_SCHEMA {
        return Err(format!("plan schema {} is not {}", head[0], PLAN_SCHEMA));
    }
    let n = head[1] as usize;
    let mut chains = Vec::with_capacity(n);
    for _ in 0..n {
        let h = take(3)?;
        let pen = h[0] as u32;
        let dot = h[1] != 0.0;
        let np = h[2] as usize;
        let mut prims = Vec::with_capacity(np);
        for _ in 0..np {
            let kind = take(1)?[0];
            let p = match kind as i32 {
                0 => {
                    let q = take(4)?;
                    Primitive::Line(Line::new(v(q[0], q[1]), v(q[2], q[3])))
                }
                1 => {
                    let q = take(5)?;
                    Primitive::Arc(Arc::new(v(q[0], q[1]), q[2], q[3], q[4]))
                }
                2 => {
                    let q = take(8)?;
                    Primitive::Cubic(Cubic::new(
                        v(q[0], q[1]),
                        v(q[2], q[3]),
                        v(q[4], q[5]),
                        v(q[6], q[7]),
                    ))
                }
                k => return Err(format!("plan buffer: unknown primitive kind {k}")),
            };
            prims.push(p);
        }
        if prims.is_empty() {
            return Err("plan buffer: a chain with no primitives".into());
        }
        chains.push(Chain { prims, dot, pen, ordered: false });
    }
    if i != buf.len() {
        return Err("plan buffer: trailing data".into());
    }
    Ok(chains)
}

/// The half-open range `[from, to)` of a plan, validated against its length.
pub fn range(chains: &[Chain], from: usize, to: usize) -> Result<&[Chain], String> {
    if from > to || to > chains.len() {
        return Err(format!(
            "plan range [{from}, {to}) is not within 0..{} in order",
            chains.len()
        ));
    }
    Ok(&chains[from..to])
}

/// Sampled representation for machine execution and estimation: chains in
/// plan order flattened at `tolerance` (mm), layout
/// `[pen, dot, n_points, x0, y0, …]` per chain. Flattening one chain never
/// depends on its neighbours, so the toolpath of a range equals the range of
/// the toolpath.
pub fn toolpath(chains: &[Chain], tolerance: f64) -> Vec<f64> {
    let mut out: Vec<f64> = Vec::new();
    let mut pts: Vec<Vec2> = Vec::new();
    for chain in chains {
        out.push(chain.pen as f64);
        out.push(if chain.dot { 1.0 } else { 0.0 });
        if chain.dot {
            let p = chain.start();
            out.push(1.0);
            out.push(p.x);
            out.push(p.y);
            continue;
        }
        // Consecutive primitives share endpoints: drop each one's duplicated
        // first point.
        let mut all: Vec<Vec2> = Vec::new();
        for prim in &chain.prims {
            pts.clear();
            prim.flatten(tolerance.max(0.01), &mut pts);
            let skip = usize::from(!all.is_empty());
            all.extend(pts.iter().skip(skip));
        }
        out.push(all.len() as f64);
        for p in &all {
            out.push(p.x);
            out.push(p.y);
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A row of exact circles: every chain starts and ends at angle 0, so all
    /// endpoints share one y. The tour's grid must not collapse on that
    /// degenerate box (it did: minutes for twelve circles).
    #[test]
    fn tour_terminates_on_collinear_closed_chains() {
        let circles: Vec<Chain> = (0..12)
            .map(|k| {
                let c = v(20.0 + k as f64 * 10.0, 30.0);
                Chain {
                    prims: vec![
                        Primitive::Arc(Arc::new(c, 8.0, 0.0, std::f64::consts::PI)),
                        Primitive::Arc(Arc::new(c, 8.0, std::f64::consts::PI, std::f64::consts::PI)),
                    ],
                    dot: false,
                    pen: 0, ordered: false,
                }
            })
            .collect();
        let t0 = std::time::Instant::now();
        let ordered = tour(circles, 200_000);
        assert_eq!(ordered.len(), 12);
        assert!(t0.elapsed().as_secs_f64() < 1.0, "tour took {:?}", t0.elapsed());
        // nearest first: the leftmost circle is closest to the origin
        assert!((ordered[0].start().x - 28.0).abs() < 1e-9);
    }

    #[test]
    fn plan_round_trips_every_primitive_kind() {
        let chains = vec![
            Chain {
                prims: vec![
                    Primitive::Line(Line::new(v(0.0, 0.0), v(1.0, 0.0))),
                    Primitive::Arc(Arc::new(v(1.0, 1.0), 1.0, -1.5707963267948966, 1.5707963267948966)),
                    Primitive::Cubic(Cubic::new(v(2.0, 1.0), v(3.0, 2.0), v(4.0, 0.0), v(5.0, 1.0))),
                ],
                dot: false,
                pen: 2, ordered: false,
            },
            Chain { prims: vec![Primitive::Line(Line::new(v(7.0, 7.0), v(7.0, 7.0)))], dot: true, pen: 0, ordered: false },
        ];
        let buf = encode_plan(&chains);
        let back = decode_plan(&buf).unwrap();
        assert_eq!(back.len(), 2);
        assert_eq!(back[0].pen, 2);
        assert!(back[1].dot);
        assert_eq!(encode_plan(&back), buf);
        assert!(decode_plan(&buf[..buf.len() - 1]).is_err());
        assert!(range(&chains, 1, 0).is_err());
        assert!(range(&chains, 0, 3).is_err());
        assert_eq!(range(&chains, 1, 2).unwrap().len(), 1);
        assert_eq!(range(&chains, 2, 2).unwrap().len(), 0);
    }
}
